demos/multi_material/compositional_buoyancy/van_keken_1997.py

```diff
@@ -0,0 +1,276 @@
+        line_string.distance(sl.Point(node_coord_x, node_coord_y)),
+    )
+    for node_coord_x, node_coord_y in zip(node_coords_x, node_coords_y)
+]
+
+# Define the signed-distance function and overwrite its value array
+signed_dist_to_interface = Function(K)
+signed_dist_to_interface.dat.data[:] = [
+    dist if is_above else -dist for is_above, dist in node_relation_to_curve
+]
+
+# Define thickness of the hyperbolic tangent profile
+min_mesh_edge_length = min(lx / nx, ly / ny)
+epsilon = Constant(min_mesh_edge_length / 4)
+
+# Initialise level set as a smooth step function
+psi.interpolate((1 + tanh(signed_dist_to_interface / 2 / epsilon)) / 2)
+# -
+
+# We next define materials present in the simulation. Here, the problem is
+# non-dimensionalised and can be described by the product of the expressions for the
+# Rayleigh and buoyancy numbers, RaB, which is also referred to as compositional
+# Rayleigh number. Therefore, we provide a value for thermal and compositional Rayleigh
+# numbers to define our approximation.
+
+# +
+buoyant_material = Material(RaB=-1)  # Vertical direction is flipped in the benchmark
+dense_material = Material(RaB=0)
+materials = [buoyant_material, dense_material]
+
+Ra = 0  # Thermal Rayleigh number
+
+RaB = field_interface(
+    [psi], [material.RaB for material in materials], method="arithmetic"
+)  # Compositional Rayleigh number, defined based on each material value and location
+
+approximation = BoussinesqApproximation(Ra, RaB=RaB)
+# -
+
+# As with the previous examples, we set up a *Timestep Adaptor* for controlling the
+# time-step length (via a CFL criterion) as the simulation advances in time. We specify
+# the initial time, initial time step $\Delta t$, and output frequency (in time units).
+
+time_now = 0  # Initial time
+delta_t = Function(R).assign(1)  # Initial time step
+output_frequency = 10  # Frequency (based on simulation time) at which to output
+t_adapt = TimestepAdaptor(
+    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
+)
+
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+Z_nullspace = create_stokes_nullspace(Z)
+
+# Boundary conditions are specified next: no slip at the top and bottom and free slip
+# on the left and ride sides.
+
+stokes_bcs = {
+    bottom_id: {"u": 0},
+    top_id: {"u": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+
+# We now set up our output. To do so, we create the output file as a ParaView Data file
+# that uses the XML-based VTK file format. We also open a file for logging and
+# instantiate G-ADOPT geodynamical diagnostic utility.
+
+# +
+output_file = VTKFile("output.pvd")
+
+plog = ParameterLog("params.log", mesh)
+plog.log_str("step time dt u_rms entrainment")
+
+gd = GeodynamicalDiagnostics(z, T, bottom_id, top_id)
+material_area = material_interface_y * lx  # Quantity of tracked material
+entrainment_height = 0.2  # Height above which entrainment diagnostic is calculated
```

Comment:
Say a little more about what this diagnostic is (i.e. the integral of a certain material above the specified height)?